File: raffia-core/src/main/java/com/codepoetics/raffia/streaming/rewriting/WeavingRewriter.java
package com.codepoetics.raffia.streaming.rewriting;

import com.codepoetics.raffia.baskets.Basket;
import com.codepoetics.raffia.operations.Updater;
import com.codepoetics.raffia.streaming.FilteringWriter;
import com.codepoetics.raffia.streaming.WeavingFilter;
import com.codepoetics.raffia.visitors.Visitors;
import com.codepoetics.raffia.writers.BasketWeavingWriter;
import com.codepoetics.raffia.writers.BasketWriter;
import com.codepoetics.raffia.writers.Writers;

class WeavingRewriter<T extends BasketWriter<T>> extends WeavingFilter<T> {

  static <T extends BasketWriter<T>> WeavingRewriter<T> weavingObject(T target, FilteringWriter<T> parent, Updater updater) {
    return weaving(target, parent, updater, Writers.weaving().beginObject());
  }

  static <T extends BasketWriter<T>> WeavingRewriter<T> weavingArray(T target, FilteringWriter<T> parent, Updater updater) {
    return weaving(target, parent, updater, Writers.weaving().beginArray());
  }

  private static <T extends BasketWriter<T>> WeavingRewriter<T> weaving(T target, FilteringWriter<T> parent, Updater updater, BasketWeavingWriter weaver) {
    return new WeavingRewriter<>(target, parent, updater, weaver);
  }

  private final T target;
  private final Updater updater;

  private WeavingRewriter(T target, FilteringWriter<T> parent, Updater updater, BasketWeavingWriter weaver) {
    super(parent, weaver);
    this.updater = updater;
    this.target = target;
  }

  @Override
  protected T writeToTarget(Basket woven) {
    return updater.update(woven).visit(Visitors.writingTo(target));
  }
}

File: raffia-core/src/main/java/com/codepoetics/raffia/streaming/rewriting/WeavingRewriter.kt
package com.codepoetics.raffia.streaming.rewriting

import com.codepoetics.raffia.baskets.Basket
import com.codepoetics.raffia.java.api.Updater
import com.codepoetics.raffia.streaming.FilteringWriter
import com.codepoetics.raffia.streaming.WeavingFilter
import com.codepoetics.raffia.writers.BasketWeavingWriter
import com.codepoetics.raffia.writers.BasketWriter
import com.codepoetics.raffia.writers.Writers
import com.codepoetics.raffia.writers.writeTo

internal class WeavingRewriter<T : BasketWriter<T>> private constructor(private val target: T, parent: FilteringWriter<T>, private val updater: Updater, weaver: BasketWeavingWriter) : WeavingFilter<T>(parent, weaver) {

    override fun writeToTarget(woven: Basket): T = updater.update(woven).writeTo(target)

    companion object {

        fun <T : BasketWriter<T>> weavingObject(
                target: T,
                parent: FilteringWriter<T>,
                updater: Updater): WeavingRewriter<T> =
                weaving(target, parent, updater, Writers.weaving().beginObject())

        fun <T : BasketWriter<T>> weavingArray(
                target: T,
                parent: FilteringWriter<T>,
                updater: Updater): WeavingRewriter<T> =
                weaving(target, parent, updater, Writers.weaving().beginArray())

        private fun <T : BasketWriter<T>> weaving(
                target: T,
                parent: FilteringWriter<T>,
                updater: Updater, weaver: BasketWeavingWriter): WeavingRewriter<T> =
                WeavingRewriter(target, parent, updater, weaver)
    }
}
